### DISCUSSION TITLE
Mixed pom-first and tycho features reactor vs. Maven Release Plugin

### DISCUSSION THREAD

**Discussion by reckart** [eclipse-tycho/tycho#940] [Q&A]
Follow up to https://github.com/eclipse/tycho/discussions/925 and https://github.com/eclipse/tycho/discussions/931.

So, following the advice from the two discussions above, I've been able to set up:

* a "traditional" update site with "features" using Tycho - for this to work, I also had to convert the "feature" modules from pom-first to tycho
* a Maven P2 update site that lives in a Maven repo, also using Tycho (not supporting "features" though)

Now, I'm trying to see how to get this to fly with the Maven Release Plugin. There is [a nice description](https://wiki.eclipse.org/Tycho/Release_Workflow) of how to use the Tycho Versions Plugin as part of the Maven Release Plugin builds to synchronize the versions in the OSGi/Eclipse metadata files with the POM versions.

However, I am having a bit of trouble with the setup...

I still have several pom-first modules in my build. These are referenced from a tycho-driven feature module in the `feature.xml`, e.g.:

```
   <plugin
         id="org.apache.uima.runtime"
         download-size="0"
         install-size="0"
         version="3.3.0.qualifier"
         unpack="false"/>
```

Now, when the `tycho-versions-plugin` scans the reactor for files it needs to update, it updates the p2 update site `category.xml` file nicely and also touches the `feature.xml` files, but it fails to update the `plugin` sections in the `feature.xml` files since these plugins are pom-first. For one, when applying a release build to a freshly checkout out project, the Eclipse/OSGi metadata files are not generated before the release plugin kicks in. 

But even in a repo that was already checked out and built once and has the generated metadata files, the `tycho-versions-plugin` fails figuring out that these `plugin` sections belong to pom-first modules from the reactor and that it should update them.

I wonder if there could be some way to allow the `tycho-versions-plugin` to recognize which pom-first maven modules will be building which bundles so that it knows to update the respective references...

**Comment by laeubi** [comment]
> * a "traditional" update site with "features" using Tycho - for this to work, I also had to convert the "feature" modules from pom-first to tycho

A update site can also contain bundles, no need for a feature at all



> Now, I'm trying to see how to get this to fly with the Maven Release Plugin. There is [a nice description](https://wiki.eclipse.org/Tycho/Release_Workflow) of how to use the Tycho Versions Plugin as part of the Maven Release Plugin builds to synchronize the versions in the OSGi/Eclipse metadata files with the POM versions.

If yo use the maven-updatesite you don't need any Tycho specific workflow that's what is is made for.

If you want to use the tycho flow it will most probably only work for a pure tycho and not for a mixed setup.

**Reply by reckart** [reply]
We're not ready at the moment to give up the features in the update site entirely. The idea at the moment is we might have two update sites for a while, one traditional with features and one that is living in Maven. 

If the maven updatesite had feature support, we'd probably directly switch to it, but we're unsure as to dropping the concept of features.

I am looking at the code of the tycho versions plugin to see if it could be extended to accommodate my use-case...

**Reply by laeubi** [reply]
Features simply do not exits at the "maven" part really, you would use a pom/bom for this in the maven world. And as you are building your feature manually tycho can't know anything about your feature so it won't be able to update any version on it.
You still can include the feature and use `dependency` instead of `include` if you just use the feature to install a set of bundles.

**Reply by reckart** [reply]
For example, let's say we'd define a well-known property like `tycho-bundle-id` that can be added to a pom-first module to mark it for consideration by the `tycho-versions-plugin` irrespective of the `packaging` and of the `artifactId`. If the `tycho-versions-plugin` sees that property, it could record a `PomVersionChange` with a target version but without a source version. When updating the `feature.xml`, that change could be processed to update all references to bundles/plugins with the ID specified in `tycho-bundle-id` to the version of that POM irrespective of what the original version in the `feature.xml` was...

**Reply by laeubi** [reply]
I don't think this will work well, even though the id *might* map the BSN this is not a requirement for a pom-first bundle.

**Reply by reckart** [reply]
Of course it is not a requirement, but it could be a way for pom-first maintainers to shoe-horn knowledge about the pom-first modules into the Tycho build. Another approach could be to add plugin a parameter like this:

```
<plugin>
  <groupId>org.eclipse.tycho</groupId>
  <artifactId>tycho-versions-plugin</artifactId>
  <configuration>
    <alwaysGeneratePomChangesForArtifacts>
      <artifact>
        <groupId>${project.groupId}</groupId>
        <artifactId>my-pom-first-module</artifactId>
        <bsn>my.pom.first.module</bsn>
      </artifact>
    </alwaysGeneratePomChangesForArtifacts >
  </configuration>
</plugin>
```

The `tycho-versions-plugin` should then update all references to the `my.pom.first.module` with the version of the `${project.groupId}:my-pom-first-module` module.

**Reply by reckart** [reply]
Just to make this clearer again:

```
<feature
      id="org.apache.uima.runtime.feature"
      label="UIMA Runtime"
      version="3.3.0.qualifier">   
  <!-- This version is updated because this feature.xml lives in a Tycho-managed module -->

   <plugin
         id="org.apache.uima.runtime"
         version="3.3.0.qualifier"/> 
  <!-- 
    This version is not updated because Tycho does not know that org.apache.uima.runtime
    lives in the pom-first reactor module uimaj-ep-runtime. If this knowledge could somehow be provided
    to the tycho-versions-plugin, it could also update this version.
  -->
</feature>

<plugin>
  <groupId>org.eclipse.tycho</groupId>
  <artifactId>tycho-versions-plugin</artifactId>
  <configuration>
    <alwaysGeneratePomChangesForArtifacts>
      <artifact>
        <groupId>${project.groupId}</groupId>
        <artifactId>uimaj-ep-runtime</artifactId>
        <bsn>org.apache.uima.runtime</bsn>
      </artifact>
    </alwaysGeneratePomChangesForArtifacts >
  </configuration>
</plugin>
```

**Reply by laeubi** [reply]
I don't know if this is working at all, if you have a version the feature it should be used as is `3.3.0.qualifier` is not a valid version here.
If you like it to use the version of the build it has to be `0.0.0` and the tycho build will replace it with the correct build version.

**Reply by reckart** [reply]
Excellent feedback! 💯

Replacing the version in the `plugin` declarations of the `feature.xml` with `0.0.0` works as expected and removes the need of having the `tycho-versions-plugin` update them at all!

I didn't try it yet, but I suppose the `version` in the `feature` element itself should remain `x.x.x.qualifier`?
